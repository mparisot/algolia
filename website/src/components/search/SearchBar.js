import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'debounce';

import { movieManager } from 'MovieManager';

import './searchBar.css';

/**
 * The search bar input
 */
export default class SearchBar extends React.Component {

    static propTypes = {
        onSearchResultsUpdated: PropTypes.func.isRequired, // called when the search results are fetched
    };

    state = {
        error: null,
    };

    onInputChange = (event) => this.doSearch(event.target.value);

    doSearch = debounce((searchedText) => {
        if(searchedText === '') return this.props.onSearchResultsUpdated([]);

        movieManager.search(searchedText).then(results => {
            this.setState({
                error: null,
            });
            this.props.onSearchResultsUpdated(results);
        }).catch(err => {
            this.setState({
                error: err.message,
            });
        });
    }, 300); // debounce to avoid too much flickering when the user is typing

    cancelSearch = () => {
        this.props.onSearchResultsUpdated([]);
    };

    render() {
        return (
            <div className="searchBar">
                <label className="searchBar-label" htmlFor="movieSearchInput">Type to search your movie</label>
                <input
                    type="search"
                    id="movieSearchInput"
                    className="searchBar-input"
                    placeholder="Type to search your movie"
                    onChange={this.onInputChange}
                />
                {this.state.error && <div className="searchBar-error">{this.state.error}</div> }
            </div>
        );
    }
}