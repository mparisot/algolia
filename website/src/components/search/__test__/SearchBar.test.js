import React from 'react';
import SearchBar from '../SearchBar';

jest.mock('debounce');
jest.mock('MovieManager');

const debounce = (fn) => fn;

test('Render the SearchBar', () => {
    const searchBar = shallow(
        <SearchBar onSearchResultsUpdated={() => null}/>,
    );

    expect(searchBar).toMatchSnapshot();
});

test('SearchBar update results when search yield results', (done) => {
    const mockSearchResultUpdate = jest.fn();

    const searchBar = mount(<SearchBar onSearchResultsUpdated={mockSearchResultUpdate}/>);

    //mockSearchResultUpdate('plop');

    const input = searchBar.find('input');

    input.simulate('focus');
    input.simulate('change', { target: { value: 'text' } });


    setTimeout(() => {
        expect(mockSearchResultUpdate).toBeCalled();
        done();
    },1);
});

test('SearchBar display error message when search fails', done => {
    const searchBar = shallow(<SearchBar onSearchResultsUpdated={(result) => null}/>);

    expect(searchBar).toMatchSnapshot();

    const input = searchBar.find('input');

    input.simulate('focus');
    input.simulate('change', { target: { value: 'error' } });

    setTimeout(() => {
        searchBar.update();
        expect(searchBar).toMatchSnapshot();
        done();
    },1);
});