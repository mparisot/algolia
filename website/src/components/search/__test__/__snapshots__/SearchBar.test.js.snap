// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the SearchBar 1`] = `
<div
  className="searchBar"
>
  <label
    className="searchBar-label"
    htmlFor="movieSearchInput"
  >
    Type to search your movie
  </label>
  <input
    className="searchBar-input"
    id="movieSearchInput"
    onChange={[Function]}
    placeholder="Type to search your movie"
    type="search"
  />
</div>
`;

exports[`SearchBar display error message when search fails 1`] = `
<div
  className="searchBar"
>
  <label
    className="searchBar-label"
    htmlFor="movieSearchInput"
  >
    Type to search your movie
  </label>
  <input
    className="searchBar-input"
    id="movieSearchInput"
    onChange={[Function]}
    placeholder="Type to search your movie"
    type="search"
  />
</div>
`;

exports[`SearchBar display error message when search fails 2`] = `
<div
  className="searchBar"
>
  <label
    className="searchBar-label"
    htmlFor="movieSearchInput"
  >
    Type to search your movie
  </label>
  <input
    className="searchBar-input"
    id="movieSearchInput"
    onChange={[Function]}
    placeholder="Type to search your movie"
    type="search"
  />
  <div
    className="searchBar-error"
  >
    Error!
  </div>
</div>
`;
